### ISSUE TITLE
Pyright doesn't know about variables defined in module if I've only imported a submodule

### ISSUE THREAD

**Issue by tomasgareau** [microsoft/pyright#2964]
**Describe the bug**
Pyright doesn't seem to know about constants defined by a module's `__init__.py` even though I've imported a submodule.

**To Reproduce**
Given the following directory structure:
```
.
├── example.py
└── module
    ├── __init__.py
    └── submodule.py
```

with the following contents:

```python
####################
# module/__init__.py
####################
MY_CONSTANT = 10


#####################
# module/submodule.py
#####################
class MySubmoduleClass:
    def __init__(self, value: int):
        self.value = value


############
# example.py
############
import module.submodule

my_instance = module.submodule.MySubmoduleClass(module.MY_CONSTANT)
print(f"my_instance has value: {my_instance.value}")
```

Pyright will warn me that "`MY_CONSTANT` is not a known member or variable":
![image](https://user-images.githubusercontent.com/10787950/151888875-4609ed78-c3a6-48c9-91c7-15814ea66a07.png)

**Expected behavior**
Pyright knows about `MY_CONSTANT` and doesn't produce any errors.

My (admittedly foggy) understanding of Python imports is that it will first run `module/__init__.py` in the process of importing `module/submodule.py`. Running the example above supports this: when running `python3 example.py`, the variable `module.MY_CONSTANT` is defined and I get the expected output "my_instance has value: 10".

**VS Code extension or command-line**
Command-line: pyright 1.1.216

**Additional context**
I'm bumping into this issue because I'm using [reorder_python_imports](https://github.com/asottile/reorder_python_imports) -- it removes duplicate imports, turning:

```
import module
import module.submodule
```

to

```
import module.submodule
```

I've got lots of ways to work around this (by e.g., wrapping the `import module` in a `if typing.TYPE_CHECKING` block, or changing my import to `from module.submodule import MySubmoduleClass`) but I was curious to know if Pyright was supposed to know about entities declared while running `module/__init__.py` in this case. No worries if you'd like to close as "intended behaviour from Pyright" in this case!

**Comment by erictraut** [comment]
This is an intended behavior.

You shouldn't rely on cross-module temporal ordering. If you want to access symbols from `module`, you should include an `import module` statement in your code. If you want to access symbols from a submodule, you should either import that submodule or the top-level module should re-export the symbol. As a static type checker, pyright enforces that. The Python runtime is very "loose" in this respect, and that allows developers to (often inadvertently) rely on cross-module implicit initialization, but this leads to code fragility. Explicit is better than implicit.
